/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

package com.liferay.commerce.product.content.web.internal.frontend.data.set.view.table;

import com.liferay.commerce.product.content.web.internal.constants.CPContentFDSNames;
import com.liferay.frontend.data.set.view.FDSView;
import com.liferay.frontend.data.set.view.table.BaseTableFDSView;
import com.liferay.frontend.data.set.view.table.FDSTableSchema;
import com.liferay.frontend.data.set.view.table.FDSTableSchemaBuilder;
import com.liferay.frontend.data.set.view.table.FDSTableSchemaBuilderFactory;
import com.liferay.portal.kernel.service.ServiceContext;
import com.liferay.portal.kernel.service.ServiceContextThreadLocal;
import com.liferay.portal.url.builder.AbsolutePortalURLBuilder;
import com.liferay.portal.url.builder.AbsolutePortalURLBuilderFactory;

import java.util.Locale;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Alessio Antonio Rendina
 */
@Component(
	property = "frontend.data.set.name=" + CPContentFDSNames.REPLACEMENT_CP_INSTANCES,
	service = FDSView.class
)
public class ReplacementCPInstanceTableFDSView extends BaseTableFDSView {

	@Override
	public FDSTableSchema getFDSTableSchema(Locale locale) {
		FDSTableSchemaBuilder fdsTableSchemaBuilder =
			_fdsTableSchemaBuilderFactory.create();

		return fdsTableSchemaBuilder.add(
			"sku", "sku",
			fdsTableSchemaField -> fdsTableSchemaField.setContentRenderer(
				"actionLink")
		).add(
			"name", "name"
		).add(
			"priceModel", "price",
			fdsTableSchemaField -> {
				ServiceContext serviceContext =
					ServiceContextThreadLocal.getServiceContext();

				AbsolutePortalURLBuilder absolutePortalURLBuilder =
					_absolutePortalURLBuilderFactory.
						getAbsolutePortalURLBuilder(
							serviceContext.getRequest());

				String moduleURL = absolutePortalURLBuilder.forESModule(
					"commerce-frontend-js", "index.js"
				).build();

				fdsTableSchemaField.setContentRendererModuleURL(
					"{PriceRenderer} from " + moduleURL);
			}
		).build();
	}

	@Reference
	private AbsolutePortalURLBuilderFactory _absolutePortalURLBuilderFactory;

	@Reference
	private FDSTableSchemaBuilderFactory _fdsTableSchemaBuilderFactory;

}